src/build/release.rs

```diff
@@ -18,45 +22,87 @@ pub fn release(p: &crate::project::Project, version: &String) -> Result<(), Erro
-        armake2::sign::cmd_keygen(PathBuf::from(&keyname))?;
-        fs::rename(format!("{}.bikey", keyname), format!("releases/keys/{}.bikey", keyname))?;
-        fs::rename(format!("{}.biprivatekey", keyname), format!("releases/keys/{}.biprivatekey", keyname))?;
+    // Generate and store key if required
+    if p.reuse_private_key.unwrap_or(false) {
+
+        // Make a new keypair if there isn't one already
+        if !Path::new(&format!("releases/keys/{}.bikey", keyname)).exists() {
+            println!("    {} {}.bikey", "KeyGen".green().bold(), keyname);
+
+            // Generate and write the keypair to disk in the current directory
+            cmd_keygen(PathBuf::from(&keyname))?;
+            fs::rename(
+                format!("{}.bikey", keyname),
+                format!("releases/keys/{}.bikey", keyname),
+            )?;
+            fs::rename(
+                format!("{}.biprivatekey", keyname),
+                format!("releases/keys/{}.biprivatekey", keyname),
+            )?;
+        }
+
+        // Read the private key from disk
+        key = BIPrivateKey::read(
+            &mut File::open(format!("releases/keys/{}.biprivatekey", keyname))
+                .expect("Failed to open private key"),
+        )
+        .expect("Failed to read private key");
+    } else {
+        // Make the private key and leave it in memory
+        key = BIPrivateKey::generate(1024, keyname.clone());
+        let public_key = key.to_public_key();
+
+        // Write the public key to disk
+        public_key.write(
+            &mut std::fs::File::create(format!("releases/keys/{}.bikey", keyname)).unwrap(),
+        )?;
     }
 
-    // Sign
-    fs::copy(format!("releases/keys/{}.bikey", keyname), format!("releases/{}/@{}/keys/{}.bikey", version, modname, keyname))?;
+    // Copy public key to specific release dir
+    fs::copy(
+        format!("releases/keys/{}.bikey", keyname),
+        format!("releases/{}/@{}/keys/{}.bikey", version, modname, keyname),
+    )?;
 
     let count = Arc::new(Mutex::new(0));
 
+    // Sign
     let mut folder = String::from("addons");
-    let dirs: Vec<_> = fs::read_dir(&folder).unwrap()
+    let dirs: Vec<_> = fs::read_dir(&folder)
+        .unwrap()
         .map(|file| file.unwrap())
         .collect();
     dirs.par_iter().for_each(|entry| {
         // TODO split copy and sign
-        if crate::build::sign::copy_sign(&folder, &entry, &p, &version).unwrap() {
+        if sign::copy_sign(&folder, &entry.path(), &p, &key).unwrap() {
             *count.lock().unwrap() += 1;
         }
     });
 
     folder = String::from("optionals");
     if Path::new(&folder).exists() {
-        if !Path::new(&format!("releases/{}/@{}/{}", version, modname, folder)).exists() {
-            fs::create_dir_all(format!("releases/{}/@{}/{}", version, modname, folder))?;
+        let addonsfolder = format!(
+            "releases/{ver}/@{mod}/{addons}", ver=version, mod=modname, addons=folder
```

Comment:
The use of variables here seems unnecessary, especially since we don't use them anywhere else. `mod` is also a keyword and shouldn't be used as a variable.

Reply:
I don't think these are variables - they just annotate the format! macro. I personally hate the empty {} ones. Happy to remove if you want.

Reply:
https://doc.rust-lang.org/std/fmt/#named-parameters

Reply:
I know they aren't variables outside the format, I just don't see a reason for their use in a simple string where each variable is in order and used once. Especially when they are using different names.

If rust ever adds string interpolation we would definitely use that, but for now, I think named parameters are not needed on simple strings like this.

Reply:
It's because the format!/printf macros are alien to me. I find named parameters much more readable. I honestly found the line much easier to understand with it in; I only understood what it did after rewriting it.

If you feel as strongly but in the other direction I'll remove it.

Reply:
#79 needs a little bit more testing, but I think that'll be a good compromise. There is no redeclaration, or like in your PR, declaring with different names. I am not a fan of changing the name for formatting, but I also think `name = name` looks ridiculous. This is the best we can do with hygienic macros unless rust one days adds string interpolation.